realm/realm-library/src/main/java/io/realm/HandlerController.java

```diff
@@ -637,9 +637,17 @@ void addToRealmResults(RealmResults<? extends RealmModel> realmResults) {
         syncRealmResults.add(realmResultsWeakReference);
     }
 
-    // add to the list of RealmObject to be notified after a commit
-    <E extends RealmModel> void addToRealmObjects(E realmobject) {
-        realmObjects.put(new WeakReference<RealmObjectProxy>((RealmObjectProxy) realmobject), null);
+    // Add to the list of RealmObject to be notified after a commit.
+    // This method will check if the object exists in the list. It won't add the same object multiple times
+    <E extends RealmObjectProxy> void addToRealmObjects(E realmObject) {
+        for (WeakReference<RealmObjectProxy> ref : realmObjects.keySet()) {
```

Comment:
The problem is `WeakReference` doesn't override `equals` and `hashcode`, so we need to extend the `WeakReference` to avoid traversing.

And also, realmObjects is not thread safe, we need to fix that. Created https://github.com/realm/realm-java/issues/2724 to track that. And this won't be fixed in this PR.